### DISCUSSION TITLE
Opting out of Astro's children optimization in MDX

### DISCUSSION THREAD

**Discussion by pomber** [withastro/roadmap#465] [Proposal]
_From [issue #6023
](https://github.com/withastro/astro/issues/6023)_

If I have this React component:

```jsx
export function LogChildren({ children }) {
  console.log({ children });
  return children;
}

```

And this MDX:

```mdx
## Hi

<LogChildren client:load>
  <a>1</a>
  <b />
</LogChildren>
```

Using Astro the component will log:

```js
{
  children: {
    '$$typeof': Symbol(react.element),
    type: [Function: StaticHtml] {
      shouldComponentUpdate: [Function (anonymous)]
    },
    key: null,
    ref: null,
    props: { value: [SlotString [HTMLString]] },
    _owner: null,
    _store: {}
  }
}
```

where `props.value` is the static HTML string `"<a>1</a><b></b>"`

While any other MDX framework logs the original `children` array:

```js
{
  children: [
    {
      '$$typeof': Symbol(react.element),
      type: 'a',
      key: null,
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    },
    {
      '$$typeof': Symbol(react.element),
      type: 'b',
      key: null,
      ref: null,
      props: {},
      _owner: null,
      _store: {}
    }
  ]
}
```

---

According to [the docs](https://docs.astro.build/en/core-concepts/framework-components/#passing-children-to-framework-components), this is by design. And it makes sense for most cases. 

But this also sacrifices some of the flexibility from MDX (see [this talk](https://www.youtube.com/watch?v=xEu3t-KJVVg)), and breaks some assumptions that other tools may rely on. 

In particular, this is the biggest issue blocking the [integration between Code Hike and Astro](https://github.com/code-hike/codehike/issues/255). For examples like [this one](https://codehike.org/demo/spotlight-preview) to work, keeping the original `children` is crucial.

I ask for a way of opting out of this optimization. I haven't used Astro too much, so maybe there's a workaround or a more idiomatic approach to issues like this one. 

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-z5pumo?file=src/content/blog/using-mdx.mdx

**Comment by bluwy** [comment]
Initially I thought this was a bug, but it seems intentional to me that it works this way, as you're not passing React elements as slots, but pure HTML elements which Astro synthesizes as a simple children with the raw string.

Supporting this would mean needing to know we're within a React/Solid/Preact component, which is tricky to do in compile time.

**Reply by bholmesdev** [reply]
Agreed. We transform hydrated client components (JSX or not) using a custom element to mount that component. This requires special logic for slot handling in particular, hence the change to `children` properties. We'd need to defer to a given framework's JSX runtime instead of our own to make this happen I think... which sort of defeats the purpose of `client` directives 😕

**Reply by matthewp** [reply]
Agreed, I think this is a result of the MDX not being React-MDX but Astro-MDX so this is the same issue that happens when you pass children to a React component inside of Astro.

Maybe there's a world where, if you're only using React as a renderer, we can transpile this differently so that instead of Astro children it is React children.

**Reply by pomber** [reply]
what about adding some way of telling Astro that the mdx file is React-MDX? to provide compatibility with other React frameworks

**Reply by bholmesdev** [reply]
@pomber This could be possible actually! Are there other motivators to support a React-only output beyond what was outlined above?

**Reply by jasikpark** [reply]
```mdx
/* jsx h */
import { h } from "React"

# hello world

I'm some React MDX instead of Astro MDX now :)
```

**Reply by bluwy** [reply]
I think to keep with Astro's nature of being framework-agnostic, each component can define how their children should be processed instead? Something like:

```mdx
# test

<Foo client:idle children:react>
  <p>I'm a React element</p>
</Foo>
```

Otherwise it would not be possible to use other framework components in that same file.

**Comment by natemoo-re** [comment]
Hello everyone—just wanted to give a quick update after spending some time investigating this.

Looks like we've found a fairly straightforward way to enable this in Astro. We're hoping to formalize our experiment with official support in the next few weeks! We'll be sure to share updates as we make progress. 

This compatibility feature will likely be exposed on the integration-level, rather than component-by-component. Exact naming is TBD, but it will probably be something like `react({ compat: true })`. There are a few trade-offs with the approach so we don't plan to enable it by default, but it will unblock Code Hike support in Astro's MDX files.

**Reply by nearestnabors** [reply]
This is great news! Thanks so much for following up!

**Reply by pomber** [reply]
Awesome! Is there any issue or PR to follow?

**Reply by natemoo-re** [reply]
Not yet, we made a [`spike/codehike`](https://github.com/withastro/astro/tree/spike/codehike) branch and we'll be sure to update here when we get a chance to implement this!

**Reply by matthewp** [reply]
We're going to release this as an experimental feature to give people a chance to try it out. Assuming it works for this need we can figure out the long-term API.

**Reply by jgoux** [reply]
Super good news, that's the only thing I'm waiting for switching from Nextra to Starlight! 😄

**Reply by pomber** [reply]
@matthewp is [`experimentalReactChildren`](https://docs.astro.build/en/guides/integrations-guide/react/#children-parsing) the feature you were mentioning? 

I [upgraded the example from the original post](https://stackblitz.com/edit/github-z5pumo-nya6at?file=package.json,astro.config.mjs,src%2Fcomponents%2FTest.jsx) and set that flag. 

This is what it logs on the server:

```
07:42:00 PM [astro] reload /src/components/Test.jsx
{
  children: [
    {
      'astro:jsx': true,
      type: 'a',
      props: [Object],
      [Symbol(astro:renderer)]: 'astro:jsx'
    },
    {
      'astro:jsx': true,
      type: 'b',
      props: {},
      [Symbol(astro:renderer)]: 'astro:jsx'
    }
  ]
}
{
  children: [
    {
      '$$typeof': Symbol(react.element),
      type: 'a',
      key: null,
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    },
    {
      '$$typeof': Symbol(react.element),
      type: 'b',
      key: null,
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    }
  ]
}
{
  children: [String (SlotString): '<a>1</a><b></b>'] [HTMLString] {
    instructions: null,
    [Symbol(astro:slot-string)]: true
  }
}
{
  children: [
    {
      '$$typeof': Symbol(react.element),
      type: 'a',
      key: null,
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    },
    {
      '$$typeof': Symbol(react.element),
      type: 'b',
      key: null,
      ref: null,
      props: [Object],
      _owner: null,
      _store: {}
    }
  ]
}
```

Looks like the component runs 4 times. 2 times renders the original React elements, 1 time renders something similar, and 1 time renders Astro's slot string as with v2.

On the client it works fine.

Is this expected?

**Reply by pomber** [reply]
Should I submit that as a bug? I'm not sure if it is expected or not. 

I'm working on Code Hike v1.0 and want to make it compatible with Astro, it will probably depend on that flag working correctly.

**Reply by matthewp** [reply]
Hey @pomber, I'll try out your example but what happens that is unwanted here? It's hard for me to understand based on logging the component.

**Reply by matthewp** [reply]
Like an example where something you want to do doesn't work due to this. I think you probably want to split apart children and put them in different places?

**Reply by pomber** [reply]
Here is an example: https://stackblitz.com/edit/github-z5pumo-df3ozr?file=src%2Fcontent%2Fblog%2Fusing-mdx.mdx

This is the MDX: 

```mdx
import { Test } from './Test.jsx';

<Test client:load>
  <div>Step 1</div>
  <div>Step 2</div>
  <div>Step 3</div>
</Test>
```

And this is the component: 

```jsx
import React from 'react';

export function Test({ children }) {
  const [index, setIndex] = React.useState(0);
  const kids = React.Children.toArray(children);
  return (
    <div>
      <div>{kids[index]}</div>
      <button onClick={() => setIndex(index + 1)}>Next</button>
    </div>
  );
}
```

This isn't working in Astro.

**Reply by matthewp** [reply]
Perfect, thank you. The thing that isn't working is it's not an array that you can grab an individual child from. I'll take a look.

**Reply by matthewp** [reply]
Ok, so this is all client-related right @pomber? AFAIK if you turn off client rendering it logs 3 children, and renders just the one. When you turn on client rendering it does the same on the server but doesn't do so in the client. Just confirming that this is the problem you are seeing.

**Reply by matthewp** [reply]
Ok, I have a POC for a fix for the client-side issue. Hoping to get something out next week.

**Reply by matthewp** [reply]
Released `@astrojs/react@3.0.4` which hopefully fixes this issue. Let me know.

**Reply by pomber** [reply]
> Released @astrojs/react@3.0.4 which hopefully fixes this issue. Let me know.

It now works, thanks!

---

I'm doing more testing, here's another failing test case:

```mdx
import { Test } from './Test.jsx';

<Test client:load>
  <div style={{ color: 'red' }}>Step 1</div>
  <div>Step 2</div>
  <div>Step 3</div>
</Test>
```

It fails with `Cannot read properties of null (reading 'useState')` but I think the real error is shown when I move the style to the second step:

```mdx
import { Test } from './Test.jsx';

<Test client:load>
  <div>Step 1</div>
  <div style={{ color: 'red' }}>Step 2</div>
  <div>Step 3</div>
</Test>
```

The error now is:
> The `style` prop expects a mapping from style properties to values, not a string. For example, style={{marginRight: spacing + 'em'}} when using JSX.

It looks like Astro is trying to assign a string to the `style` prop of a React element.

Repro: https://stackblitz.com/edit/github-z5pumo-o9qmtv?file=src%2Fcontent%2Fblog%2Fusing-mdx.mdx

**Reply by bholmesdev** [reply]
Ah yikes! Mind copying that into an issue for us to triage?

**Reply by matthewp** [reply]
This one I don't think we can do anything about. In Astro an MDX file gets compiled to an Astro component. So syntactically it works like in `.astro` files. And in Astro children become strings that are handed to the parent component.

The workaround that we implemented with `experimentalReactChildren` was to convert said string back to a React VNode before running the component. It allows you to manipulation children, but it doesn't make it so that the MDX file has React semantics.

If you're converting your site and expecting a 1:1 conversion, that's not possible. It will require some refactor and doing things a little differently.

I think what you're probably really wanting is for MDX to be compiled to React. Which would be cool, but would require that we first support React as page components, not something we currently allow.

**Reply by matthewp** [reply]
I started a discussion on this topic, https://github.com/withastro/roadmap/discussions/750

**Reply by StandardGage** [reply]
Hey hoping this is the correct place for this. While trying to build a carousel I found that the `experimentalReactChildren` flag has issues when passing any sort of image to a React component, specifically when the `client` directive is used. I've tested with Astro's Image component, as well as the basic `<img>` element. The image will appear for a split second before disappearing. Looking at the inspect console shows this error `Warning: An error occurred during hydration. The server HTML was replaced with client content in <astro-island>.` So it seems Astro is replacing the content. The component works perfectly if hydration is not used, but this kind of defeats the purpose of using react here.

Here is the Carousel component:
```js
import React from "react";

interface CarouselProps {
  children: React.ReactNode;
}

export default function Carousel({ children }: CarouselProps) {
  return <div>{children}</div>;
}
```

And here is the implementation of it in index.astro
```js
<Carousel client:load>
			<Image src='https://picsum.photos/200' alt="image" width='200' height='200'/>
			<Image src='https://picsum.photos/200' alt="image" width='200' height='200'/>
			<Image src={image} alt="image" width='200' height='200'/>
</Carousel>
```

And finally here is a code [sandbox](https://codesandbox.io/p/devbox/astro-image-as-react-children-sysvvn?embed=1&file=%2Fsrc%2Fpages%2Findex.astro&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrfpm14k00072a66bsg2pn29%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrfpm14k00022a66hlr2pewl%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrfpm14k00042a66tdrat1xk%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrfpm14k00062a66nu8jzc6c%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B62.34511087824278%252C37.65488912175722%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrfpm14k00022a66hlr2pewl%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrfpm14k00012a66hakexvxe%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clrfpm14k00022a66hlr2pewl%2522%252C%2522activeTabId%2522%253A%2522clrfpm14k00012a66hakexvxe%2522%257D%252C%2522clrfpm14k00062a66nu8jzc6c%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrfpm14k00052a66hxkdhcok%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A4321%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clrfpm14k00062a66nu8jzc6c%2522%252C%2522activeTabId%2522%253A%2522clrfpm14k00052a66hxkdhcok%2522%257D%252C%2522clrfpm14k00042a66tdrat1xk%2522%253A%257B%2522id%2522%253A%2522clrfpm14k00042a66tdrat1xk%2522%252C%2522activeTabId%2522%253A%2522clrfpo4pi00bv2a66k0bi5mun%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrfpm14k00032a660eopvxdb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clrfpo4pi00bv2a66k0bi5mun%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clrfpo4dh000edxfmagui9ahn%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%252C%2522id%2522%253A%2522clrfpp43300f22a662spwk92a%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) that I tested with.

**Reply by bholmesdev** [reply]
Thanks @StandardGage! I'd report that as an issue on https://github.com/withastro/astro . This doesn't sound intentional

**Reply by matthewp** [reply]
@StandardGage that flag only works in MDX files.

**Reply by aloTu** [reply]
Web Component with children don't works too.

**Reply by StandardGage** [reply]
@matthewp Could you point me in the direction of more information about this. In the docs I don't see any mention of MDX.
https://docs.astro.build/en/guides/integrations-guide/react/#children-parsing

**Reply by tomByrer** [reply]
@pomber @natemoo-re Have you guys tried using codehike since [astrojs/mdx 3.0
](https://github.com/withastro/astro/blob/refs/heads/main/packages/integrations/mdx/CHANGELOG.md)?  Namely since these new changes:
>Refactors the MDX transformation to rely only on the unified pipeline. Babel and esbuild transformations are removed, which should result in faster build times. The refactor requires using Astro v4.8.0 but no other changes are necessary... Allows integrations after the MDX integration to update markdown.remarkPlugins and markdown.rehypePlugins, and have the plugins work in MDX too.

I've been trying to add in codehike v1-alpha, & unsure about the errors.

**Reply by pomber** [reply]
> I've been trying to add in codehike v1-alpha, & unsure about the errors

haven't tried but for what I understand it won't work (mostly because codehike passes React elements inside props and Astro needs slots for that)

I have an idea for an adapter, but will have to wait until we ship v1. I'm also open to collaborate with the Astro team to make something more idiomatic that integrates better with content collections.

**Comment by effulgentsia** [comment]
Does [experimentalReactChildren](https://docs.astro.build/en/guides/integrations-guide/react/#children-parsing) resolve the original problem that sparked this discussion, or is there something else being discussed here?

**Reply by delucis** [reply]
Yeah, that’s the option introduced with this issue in mind. (Not sure whether that resolved things or not.)

**Reply by aleclarson** [reply]
It'd be great to see Preact support for this too!

**Reply by matthewp** [reply]
This has introduced a lot of bugs. I'm not sure if it will ever become stable.

**Comment by tisonkun** [comment]
I found https://www.mintlify.com/blog/astro-react-children can help.

**Reply by tisonkun** [reply]
... and thus this file https://github.com/mintlify/astro/blob/362922e7src/utils/extract-react-components.ts#L285
